File: src/com/train/algorithm/tree/implementInC++/PathSumIII .cpp
//
// Created by 罗斯 on 1/11/2022.
//
#include <unordered_map>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
//        if (!root) return 0;
//        return numberOfPaths(root, sum) + pathSum(root->left, sum) + pathSum(root->right, sum);

        countPathSum(root, targetSum, 0);
        return count;
    }
private:
    unordered_map<long, int> map;
    int count = 0;

    int numberOfPaths(TreeNode* root, int left) {
        if (!root) return 0;
        left -= root->val;
        return (left == 0 ? 1 : 0) + numberOfPaths(root->left, left) + numberOfPaths(root->right, left);
    }

    void countPathSum(TreeNode* root, int target, long sum){
        if(!root)
            return;
        sum += root->val;
        if(sum == target)
            count++;
        if(map.find(sum - target) != map.end())
            count += map[sum - target];
        map[sum]++;
        countPathSum(root->left, target, sum);
        countPathSum(root->right, target, sum);
        map[sum]--;
    }
};

File: src/com/train/algorithm/tree/implementInPython/PathSumIII.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def pathSum(self, root, sum):
        global result
        result = 0

        def dfs(node, target):
            if node is None: return
            find_path_from_node(node, target)
            dfs(node.left, target)
            dfs(node.right, target)

        def find_path_from_node(node, target):
            global result
            if node is None: return
            if node.val == target: result += 1
            find_path_from_node(node.left, target - node.val)
            find_path_from_node(node.right, target - node.val)

        dfs(root, sum)

        return result